src/trading/NewOrders.sol

```diff
@@ -0,0 +1,338 @@
+        require(_outcome < _market.getNumberOfOutcomes());
+        _orderId = getOrderIdHash(_type, _market, _fxpAmount, _fxpPrice, _sender, block.number, _outcome, _fxpMoneyEscrowed, _fxpSharesEscrowed);
+        insertOrderIntoList(_orderId, _type, _market, _outcome, _fxpPrice, _betterOrderId, _worseOrderId);
+        orders[_orderId].fxpPrice = _fxpPrice;
+        orders[_orderId].fxpAmount = _fxpAmount;
+        orders[_orderId].owner = _sender;
+        orders[_orderId].fxpMoneyEscrowed = _fxpMoneyEscrowed;
+        orders[_orderId].fxpSharesEscrowed = _fxpSharesEscrowed;
+        MakeOrder(_market, _sender, _type, _fxpPrice, _fxpAmount, _outcome, _orderId, _fxpMoneyEscrowed, _fxpSharesEscrowed, _tradeGroupId);
+        return _orderId;
+    }
+
+    function removeOrder(bytes20 _orderId, Trading.TradeTypes _type, IMarket _market, uint8 _outcome) public onlyWhitelistedCallers returns (bool) { 
+        require(tx.gasprice <= orders[_orderId].gasPrice);
+        removeOrderFromList(_orderId, _type, _market, _outcome);
+        orders[_orderId].fxpPrice = 0;
+        orders[_orderId].fxpAmount = 0;
+        orders[_orderId].owner = 0;
+        orders[_orderId].fxpMoneyEscrowed = 0;
+        orders[_orderId].fxpSharesEscrowed = 0;
+        orders[_orderId].gasPrice = 0;
+        return true;
+    }
+
+    function fillOrder(bytes20 _orderId, Trading.TradeTypes _orderType, IMarket _market, uint8 _orderOutcome, uint256 _sharesFilled, uint256 _tokensFilled) public onlyWhitelistedCallers returns (bool) { 
+        // FIXME: Should eventually be changed to `require(_market.getTypeName() == "Market")`
+        require(_market != address(0));
+        require(_orderOutcome < _market.getNumberOfOutcomes());
+        require(_orderType == Trading.TradeTypes.Bid || _orderType == Trading.TradeTypes.Ask);
+        require(_orderId != 0);
+        require(_sharesFilled <= orders[_orderId].fxpSharesEscrowed);
+        require(_tokensFilled <= orders[_orderId].fxpMoneyEscrowed);
+        require(tx.gasprice <= orders[_orderId].gasPrice);
+        require(orders[_orderId].fxpPrice <= _market.getMaxDisplayPrice());
+        require(orders[_orderId].fxpPrice >= _market.getMinDisplayPrice());
+        require(_market.getMaxDisplayPrice() + _market.getMinDisplayPrice() <= 2**254);
+        uint256 _fill = 0;
+        if (_orderType == Trading.TradeTypes.Bid) {
+            // We can't use safeSub here because it disallows subtracting negative numbers. Worst case here is an operation of 2**254 - 1 as required above, which won't overflow
+            _fill = _sharesFilled + (_tokensFilled.fxpDiv(orders[_orderId].fxpPrice - _market.getMinDisplayPrice()));
+        }
+        if (_orderType == Trading.TradeTypes.Ask) {
+            // We can't use safeSub here because it disallows subtracting negative numbers. Worst case here is an operation of 2**254 - 1 as required above, which won't overflow
```

Comment:
At some point (doesn't have to be in this PR) we should add a `safeMathInt256` (if we haven't already) that understands how to safely subtract and add negative numbers.  Then we can delete this comment and use the safe math functions for `int256`.

For now, feel free to just update the comment to be a TODO indicating that we should make this use a safe math library that supports int256.

Reply:
I recently created a SafeMathInt256.  I'll update this line to use it.